README.md
logworks is a convenience wrapper for the logging Python module. It is useful for me and the work I do at IHCantabria. 

I make it public with best wishes, but no implication is made about its potential utility and usability for others. logworks is distributed as free software, under the GPLv3.


## Some examples

NOTE: GitHub's markdown parser discards some HTML. As a result, the outputs below lack color.

### Default

```python
from logworks import logworks

logger = logworks.ConsoleLogger()

logger.debug("Verbose debug")
logger.info("This is some info")
logger.ok("Everything is ok")
logger.warning("Danger! Danger!")
logger.error("Something went wrong")
```
    
Yields the following (your exact colors may vary). Note debug text is not logged.

<pre>
2018-04-11 12:10:35 <span style="color: blue">[INFO]</span> This is some info
2018-04-11 12:10:35 <span style="color: green">[OK]</span> Everything is ok
2018-04-11 12:10:35 <span style="color: goldenrod">[WARNING]</span> Danger! Danger!
2018-04-11 12:10:35 <span style="color: red">[ERROR]</span> Something went wrong
</pre>

### Debug

```python
import logging
from logworks import logworks

logger = logworks.ConsoleLogger(level=logging.DEBUG)

logger.debug("Verbose debug")
```
    
Yields (now it logs debug). Note that default color for debug is no color.

<pre>
2018-04-11 12:10:35 [DEBUG] Verbose debug
</pre>

### No colors

```python
from logworks import logworks

logger = logworks.ConsoleLogger(use_color=False)

logger.info("This is some info")
logger.ok("Everything is ok")
logger.warning("Danger! Danger!")
logger.error("Something went wrong")
```
    
Yields:

<pre>
2018-04-11 12:10:35 [INFO] This is some info
2018-04-11 12:10:35 [OK] Everything is ok
2018-04-11 12:10:35 [WARNING] Danger! Danger!
2018-04-11 12:10:35 [ERROR] Something went wrong
</pre>

### Custom formatter

```python
from logworks import logworks

myformatter = logworks.get_formatter(
    format='{clevelname} - {asctime} - {message}',
    date_format="%H:%M:%S"
)

logger = logworks.ConsoleLogger(console_formatter=myformatter)

logger.info("This is some custom info")
```
    
Yields:

<pre>
<span style="color: blue">[INFO]</span> - 12:10:35 - This is some custom info
</pre>


CHANGELOG.txt
--- TODO

  - Be able to capture logged string in tests

--- current
  
  - 100% test coverage
  - Added CHANGELOG.txt (this file)
  - Ignored test_version_ok() test (which should and does fail)
  - Provided all functionality from logworks, do not require import logging

--- v0.7.5 [2018.05.22]

  - Multiple bugfixes for PyPI

--- v0.7 [2018.05.22]

  - First version in PyPI


test/test.sh
# Run tests and then show coverage report (only if no errors):
python -m coverage run --source="." -m unittest discover -s test -v || exit

python -m coverage report -m


test/test_logworks.py
# Standard libs:
import os
import mock
import logging
import unittest
from io import StringIO

# Our libs:
from src import logworks


# Classes:
class TestLogger(unittest.TestCase):
    """Test Logger() class."""

    # Setup and teardown:
    def setUp(self):
        self.TEXTS = ["something", "yadda yadda", "", None]
        self.COLORS = {
            "name": 31,
            "info": 32,
            "warning": 33,
            "error": 34,
        }
        with mock.patch("sys.stdout"), mock.patch("sys.stderr"):
            self.logger = logworks.Logger()

    def tearDown(self):
        self.logger = None
        if os.path.isfile("logworks.log"):
            os.unlink("logworks.log")

    # Test constructor:
    def test_constructor_no_args(self):
        # Run:
        logger = self.logger

        # Assert:
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.no_color, False)
        self.assertIsInstance(logger.logger, logging.Logger)
    
    def test_constructor_no_color(self):
        # Run:
        with mock.patch("sys.stdout", new_callable=StringIO) as stdout, mock.patch("sys.stderr"):
            logger = logworks.Logger(use_color=False)

        # Assert:
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.no_color, True)
        self.assertIsInstance(logger.logger, logging.Logger)
    
    def test_constructor_with_empty_conf_file(self):
        # Run:
        with mock.patch("sys.stdout", new_callable=StringIO) as stdout, mock.patch("sys.stderr"):
            logger = logworks.Logger(conf_fn="whatever")

        # Assert:
        self.assertIsInstance(stdout.getvalue(), str)
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.conf, {})
        self.assertIsInstance(logger.logger, logging.Logger)
    
    # Test print shortcuts:
    def test_info(self):
        # Prepare:
        logger = self.logger
        logger.logger.info = mock.Mock()
        logger.with_info_color = mock.Mock()

        for text in self.TEXTS:
            # Prepare:
            logger.logger.info.reset_mock()
            logger.with_info_color.reset_mock()

            # Run:
            logger.info(text)

            # Assert:
            logger.logger.info.assert_called_once()
            self.assertIn(text, logger.logger.info.call_args[0])
            logger.with_info_color.assert_called_once()

    def test_warning(self):
        # Prepare:
        logger = self.logger
        logger.logger.warning = mock.Mock()
        logger.with_warning_color = mock.Mock()

        for text in self.TEXTS:
            # Prepare:
            logger.logger.warning.reset_mock()
            logger.with_warning_color.reset_mock()

            # Run:
            logger.warning(text)

            # Assert:
            logger.logger.warning.assert_called_once()
            self.assertIn(text, logger.logger.warning.call_args[0])
            logger.with_warning_color.assert_called_once()

    def test_error(self):
        # Prepare:
        logger = self.logger
        logger.logger.error = mock.Mock()
        logger.with_error_color = mock.Mock()

        for text in self.TEXTS:
            # Prepare:
            logger.logger.error.reset_mock()
            logger.with_error_color.reset_mock()

            # Run:
            logger.error(text)

            # Assert:
            logger.logger.error.assert_called_once()
            self.assertIn(text, logger.logger.error.call_args[0])
            logger.with_error_color.assert_called_once()

    def test_debug(self):
        # Prepare:
        logger = self.logger
        logger.logger.debug = mock.Mock()
        logger.with_debug_color = mock.Mock()

        for text in self.TEXTS:
            # Prepare:
            logger.logger.debug.reset_mock()
            logger.with_debug_color.reset_mock()

            # Run:
            logger.debug(text)

            # Assert:
            logger.logger.debug.assert_called_once()
            self.assertIn(text, logger.logger.debug.call_args[0])
            logger.with_debug_color.assert_called_once()

    def test_ok(self):
        # Prepare:
        logger = self.logger
        logger.logger.info = mock.Mock()  # yes, "info". This is correct.
        logger.with_ok_color = mock.Mock()

        for text in self.TEXTS:
            # Prepare:
            logger.logger.info.reset_mock()
            logger.with_ok_color.reset_mock()

            # Run:
            logger.ok(text)

            # Assert:
            logger.logger.info.assert_called_once()
            self.assertIn(text, logger.logger.info.call_args[0])
            logger.with_ok_color.assert_called_once()

    # Test colorizers:
    def test_with_name_color(self):
        # Prepare:
        logger = self.logger

        # Run:
        for text in self.TEXTS:
            ret = logger.with_name_color(text)

            # Assert:
            self.assertIn(str(text), str(ret))

    def test_with_info_color(self):
        # Prepare:
        logger = self.logger

        # Run:
        for text in self.TEXTS:
            ret = logger.with_info_color(text)

            # Assert:
            self.assertIn(str(text), str(ret))

    def test_with_warning_color(self):
        # Prepare:
        logger = self.logger

        # Run:
        for text in self.TEXTS:
            ret = logger.with_warning_color(text)

            # Assert:
            self.assertIn(str(text), str(ret))

    def test_with_error_color(self):
        # Prepare:
        logger = self.logger

        # Run:
        for text in self.TEXTS:
            ret = logger.with_error_color(text)

            # Assert:
            self.assertIn(str(text), str(ret))

    def test_with_debug_color(self):
        # Prepare:
        logger = self.logger

        # Run:
        for text in self.TEXTS:
            ret = logger.with_debug_color(text)

            # Assert:
            self.assertIn(str(text), str(ret))

    def test_with_ok_color(self):
        # Prepare:
        logger = self.logger

        # Run:
        for text in self.TEXTS:
            ret = logger.with_ok_color(text)

            # Assert:
            self.assertIn(str(text), str(ret))

    def test_colorize_as_with_colors(self):
        # Prepare:
        logger = self.logger
        with mock.patch("src.logworks.Logger.use_colors"): # to be able to override it (being a @property)
            logger.use_colors = True
            logger.conf = {"colors": self.COLORS}

            # Run:
            for which in self.COLORS:
                ret = logger._colorize_as(which, which)

                # Assert:
                self.assertIn(which, ret)
                self.assertNotEqual(which, ret)

    def test_colorize_as_without_colors(self):
        # Prepare:
        logger = self.logger
        with mock.patch("src.logworks.Logger.use_colors"): # to be able to override it (being a @property)
            logger.use_colors = False
            logger.conf = {"colors": self.COLORS}

            # Run:
            for which in self.COLORS:
                ret = logger._colorize_as(which, which)

                # Assert:
                self.assertEqual(which, ret)

    def test_color_for_can(self):
        # Prepare:
        logger = self.logger
        logger.conf = {"colors": self.COLORS}

        for which in self.COLORS:
            # Run:
            ret = logger._color_for(which)

            # Assert:
            self.assertEqual(ret, self.COLORS[which])

    def test_color_for_cant(self):
        # Prepare:
        logger = self.logger
        logger.conf = {}

        for which in self.COLORS:
            # Run:
            ret = logger._color_for(which)

            # Assert:
            self.assertEqual(ret, 0)

    def test_colorize(self):
        # Run:
        for text in self.TEXTS:
            ret = logworks.Logger.colorize(text, 666)
            
            # Assert:
            self.assertIn(str(text), str(ret))
            self.assertIn("666", str(ret))

    def test_colorize_none(self):
        # Run:
        for text in self.TEXTS:
            ret = logworks.Logger.colorize(text, None)
            
            # Assert:
            self.assertEqual(ret, text)

    # Test other:
    def test_use_colors_no_conf(self):
        # Prepare:
        logger = self.logger

        # Assert:
        self.assertTrue(logger.use_colors)

    def test_use_colors_empty_conf(self):
        # Prepare:
        logger = self.logger
        logger.conf = {}

        # Assert:
        self.assertFalse(logger.use_colors)

    def test_use_colors_nocolor_overrides(self):
        # Prepare:
        logger = self.logger
        logger.no_color = True
        logger.conf = {"colorize": True}

        # Assert:
        self.assertFalse(logger.use_colors)

    def test_use_colors_yes(self):
        # Prepare:
        logger = self.logger
        logger.nocolor = False
        logger.conf = {"colorize": True}

        # Assert:
        self.assertTrue(logger.use_colors)

    def test_read_conf(self):
        # Prepare:
        conf_data = {"a": 3}
        conf_string = '{"a": 3}'
        fn = "file_which_does_not_exist"

        # Run:
        with mock.patch("builtins.open", mock.mock_open(read_data=conf_string)) as mock_file:
            ret = logworks.Logger.read_conf(fn)
            mock_file.assert_called_with(fn)
            
        # Assert:
        self.assertEqual(ret, conf_data)

    def test_read_conf_no_conf(self):
        # Run:
        ret = logworks.Logger.read_conf()

        # Assert:
        self.assertEqual(ret, {})

    def test_read_conf_cant_read(self):
        with mock.patch("sys.stdout"):
            # Run:
            ret = logworks.Logger.read_conf("file_which_does_not_exist")
            
        # Assert:
        self.assertEqual(ret, {})


class TestConsoleLogger(unittest.TestCase):
    """Test ConsoleLogger() class."""

    # Setup and teardown:
    def setUp(self):
        self.TEXTS = ["something", "yadda yadda", "", None]
        self.COLORS = {
            "name": 31,
            "info": 32,
            "warning": 33,
            "error": 34,
        }
        with mock.patch("sys.stdout"), mock.patch("sys.stderr"):
            self.logger = logworks.ConsoleLogger()

    def tearDown(self):
        pass

    # Test constructor:
    def test_constructor_no_args(self):
        # Run:
        logger = self.logger

        # Assert:
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.no_color, False)
        self.assertIsInstance(logger.logger, logging.Logger)
    
    def test_constructor_no_color(self):
        # Run:
        with mock.patch("sys.stdout"), mock.patch("sys.stderr"):
            logger = logworks.ConsoleLogger(use_color=False)

        # Assert:
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.no_color, True)
        self.assertIsInstance(logger.logger, logging.Logger)
    
    def test_constructor_with_empty_conf_file(self):
        # Run:
        with mock.patch("sys.stdout", new_callable=StringIO) as stdout, mock.patch("sys.stderr"):
            logger = logworks.ConsoleLogger(conf_fn="whatever")

        # Assert:
        self.assertIsInstance(stdout.getvalue(), str)
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.conf, {})
        self.assertIsInstance(logger.logger, logging.Logger)


class TestFileLogger(unittest.TestCase):
    """Test ConsoleLogger() class."""

    # Setup and teardown:
    def setUp(self):
        self.TEXTS = ["something", "yadda yadda", "", None]
        self.COLORS = {
            "name": 31,
            "info": 32,
            "warning": 33,
            "error": 34,
        }
        with mock.patch("sys.stdout"), mock.patch("sys.stderr"):
            self.logger = logworks.FileLogger()

    def tearDown(self):
        self.logger = None
        if os.path.isfile("logworks.log"):
            os.unlink("logworks.log")

    # Test constructor:
    def test_constructor_no_args(self):
        # Run:
        logger = self.logger

        # Assert:
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.no_color, True)
        self.assertIsInstance(logger.logger, logging.Logger)
    
    def test_constructor_no_color(self):
        # Run:
        logger = self.logger

        # Assert:
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.no_color, True)
        self.assertIsInstance(logger.logger, logging.Logger)
    
    def test_constructor_with_empty_conf_file(self):
        # Run:
        with mock.patch("sys.stdout", new_callable=StringIO) as stdout, mock.patch("sys.stderr"):
            logger = logworks.ConsoleLogger(conf_fn="whatever")

        # Assert:
        self.assertIsInstance(stdout.getvalue(), str)
        self.assertIsInstance(logger.conf, dict)
        self.assertEqual(logger.conf, {})
        self.assertIsInstance(logger.logger, logging.Logger)


class TestMain(unittest.TestCase):
    """Test stuff outside Logger() class."""

    # Setup and teardown:
    def setUp(self):
        pass

    def tearDown(self):
        if os.path.isfile("logworks.log"):
            os.unlink("logworks.log")

    # Test version:
    @unittest.skip("is this test impossible?")
    def test_version_ok(self):
        self.assertIsInstance(logworks.__version__, str)

    def test_version_ko(self):
        # Prepare:
        import importlib
        import pkg_resources
        
        # Run:
        with mock.patch("pkg_resources.get_distribution", side_effect=pkg_resources.DistributionNotFound):
            importlib.reload(logworks)
        
        # Assert:
        self.assertIsNone(logworks.__version__)

        # Clean:
        importlib.reload(logworks)

    # Other:
    def test_get_formatter_default(self):
        # Run:
        fmt = logworks.get_formatter()

        # Assert:
        self.assertIsInstance(fmt, logging.Formatter)

    @unittest.skip("right now, this is useless")
    def test_get_formatter_custom_format(self):
        # We use a specific name for logger, to override logger(s) generated in previous tests:
        fmt = logworks.get_formatter(format='{clevelname} - {message}')
        logger = logworks.ConsoleLogger(
            which_logger="test_get_formatter_custom_format",
            console_formatter=fmt,
        )
        ###logger.info("post")

        # Assert:
        self.assertIsInstance(fmt, logging.Formatter)


src/logworks.py
# Standard libs:
import json
import logging
import pkg_resources

# Version:
try:
    __version__ = pkg_resources.get_distribution("logworks").version
except pkg_resources.DistributionNotFound:  # pkg not installed
    __version__ = None

# Globals:
DEFAULT_CONSOLE_FORMATTER = logging.Formatter(
    fmt='{asctime} {clevelname} {message}',
    datefmt="%Y-%m-%d %H:%M:%S",
    style="{"
)
DEFAULT_FILE_FORMATTER = logging.Formatter(
    fmt='{asctime} [{levelname}] {message}',
    datefmt="%Y-%m-%d %H:%M:%S",
    style="{"
)
DEFAULT_CONF = {
    "colorize": True,
    "logfile": "logworks.log",
    "colors": {
        "debug": 37,
        "info": 34,
        "warning": 33,
        "error": 31,
        "name": 36,
        "ok": 32,
    }
}


# Functions:
def get_formatter(format='{asctime} {clevelname} {message}', date_format="%Y-%m-%d %H:%M:%S"):
    """Helper to produce a custom logging format."""

    return logging.Formatter(
        fmt=format,
        datefmt=date_format,
        style="{"
    )


# Classes:
class Logger(object):
    """Class to hold logging stuff."""
    
    # Constructor:
    def __init__(self, conf_fn=None, use_color=True, console_formatter=DEFAULT_CONSOLE_FORMATTER,
                 file_formatter=DEFAULT_FILE_FORMATTER, which_logger=__name__, level=logging.INFO,
                 console_output=True, file_output=True, logfile=None):
        # If given a configuration file name, try to read it:
        if conf_fn:
            self.conf = Logger.read_conf(conf_fn)
        else:
            self.conf = DEFAULT_CONF

        # Avoid colors?:
        self.no_color = not use_color

        # Logger object:
        self.logger = logging.getLogger(which_logger)
        self.logger.setLevel(level)

        # Console output handler:
        if console_output:
            ch = logging.StreamHandler()
            ch.setFormatter(console_formatter)
            ch.setLevel(level)
            self.logger.addHandler(ch)

        # File output handler:
        if file_output:
            if not logfile:
                logfile = self.conf.get("logfile", None)

            if logfile:
                fh = logging.FileHandler(logfile)
                fh.setFormatter(file_formatter)
                self.logger.addHandler(fh)

    # Public methods:
    def debug(self, text):
        """Log (print) 'text' as debug."""

        extra = {
            "clevelname": self.with_debug_color("[DEBUG]"),
        }

        self.logger.debug(text, extra=extra)

    def info(self, text):
        """Log (print) 'text' as info."""

        extra = {
            "clevelname": self.with_info_color("[INFO]"),
        }

        self.logger.info(text, extra=extra)

    def ok(self, text):
        """Log (print) 'text' as OK."""

        extra = {
            "clevelname": self.with_ok_color("[OK]"),
        }

        self.logger.info(text, extra=extra)

    def warning(self, text):
        """Log (print) 'text' as warning."""

        extra = {
            "clevelname": self.with_warning_color("[WARNING]"),
        }

        self.logger.warning(text, extra=extra)

    def error(self, text):
        """Log (print) 'text' as error."""

        extra = {
            "clevelname": self.with_error_color("[ERROR]"),
        }

        self.logger.error(text, extra=extra)

    def with_debug_color(self, text):
        """Return 'text' with color for name."""

        return self._colorize_as(text, "debug")

    def with_name_color(self, text):
        """Return 'text' with color for name."""

        return self._colorize_as(text, "name")

    def with_info_color(self, text):
        """Return 'text' with color for 'info'."""

        return self._colorize_as(text, "info")

    def with_ok_color(self, text):
        """Return 'text' with color for 'ok'."""

        return self._colorize_as(text, "ok")

    def with_error_color(self, text):
        """Return 'text' with color for 'error'."""

        return self._colorize_as(text, "error")

    def with_warning_color(self, text):
        """Return 'text' with color for 'warning'."""

        return self._colorize_as(text, "warning")

    # Private methods:
    def _colorize_as(self, text, which):
        """Return 'text' with color for 'which' type of text."""

        if self.use_colors:
            return Logger.colorize(text, self._color_for(which))
        
        return text

    def _color_for(self, which):
        """Return color for 'which'."""

        try:
            return self.conf["colors"][which]
        except:
            return 0

    # Public properties:
    @property
    def use_colors(self):
        """Return True if colors should be used in terminal.
        False otherwise.
        """
        if self.no_color:
            return False

        return "colorize" in self.conf and self.conf["colorize"]

    # Static methods:
    @staticmethod
    def read_conf(fn=None):
        """Read configuration file 'fn' and return dictionary with configuration.
        Return empty dir if we couldn't read.
        """
        if not fn:
            return {}

        try:
            with open(fn) as f_conf:
                return json.load(f_conf)
        except FileNotFoundError:
            print("Could not read logger configuration file '{f}'. Ignoring...".format(f=fn))
            return {}

    @staticmethod
    def colorize(text, color_number=None):
        """Return colorized version of 'text', with terminal color 'color_number' (31, 32...).
        Return bare text if 'color_number' is None.
        """
        if color_number is None:
            return text

        return "\033[{n}m{t}\033[0m".format(t=text, n=color_number)


class ConsoleLogger(Logger):
    """A Logger() for console output only."""

    # Constructor:
    def __init__(self,
                 logfile=None,
                 conf_fn=None,
                 console_formatter=DEFAULT_CONSOLE_FORMATTER,
                 which_logger=__name__,
                 use_color=True,
                 level=logging.DEBUG):
        super().__init__(
                conf_fn=conf_fn,
                console_formatter=console_formatter,
                which_logger=which_logger,
                level=level,
                console_output=True,
                use_color=use_color,
                file_output=False,
                logfile=logfile)


class FileLogger(Logger):
    """A Logger() for file output only."""

    # Constructor:
    def __init__(self,
                 logfile=None,
                 conf_fn=None,
                 file_formatter=DEFAULT_FILE_FORMATTER,
                 which_logger=__name__,
                 level=logging.DEBUG):
        super().__init__(
                conf_fn=conf_fn,
                file_formatter=file_formatter,
                which_logger=which_logger,
                level=level,
                console_output=False,
                file_output=True,
                use_color=False,
                logfile=logfile)

